File: Lovecraft/Api/Controllers/StoryController.cs
using System.Linq.Expressions;
using System.Security.Claims;
using Azure.Core;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;
using Lovecraft.Api.Helper;
using Lovecraft.Api.Model;
using Lovecraft.Api.Model.PublicApi;
using Lovecraft.Api.Repository;
using Lovecraft.Model;
using Lovecraft.Model.PublicApi;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json;

namespace Lovecraft.Api.Controllers
{
	[Route("stories")]
	[ApiController]
	public class StoryController : ControllerBase
	{
		public IConfiguration _configuration;
		private readonly ICommonRepository<Story> _storyRepository;
		private readonly ICommonRepository<Page> _pageRepository;
		private readonly ICommonRepository<Event> _eventRepository;
		private readonly ICommonRepository<StoryStoryTag> _storyStoryTagRepository;

		private static readonly JsonSerializerSettings MultiPartMessageJsonSerializerSettings = new JsonSerializerSettings
		{
			NullValueHandling = NullValueHandling.Ignore
		};
		public StoryController(ICommonRepository<Story> storyRepository, ICommonRepository<Page> pageRepository, IConfiguration configuration, ICommonRepository<Event> eventRepository, ICommonRepository<StoryStoryTag> storyStoryTagRepository)
		{
			_storyRepository = storyRepository;
			_pageRepository = pageRepository;
			_eventRepository = eventRepository;
			_configuration = configuration;
			_storyStoryTagRepository = storyStoryTagRepository;
		}

		[HttpGet]
		public ActionResult GetAll([FromQuery] int? teamsId,[FromQuery]int? eventId, [FromQuery] string? search, [FromQuery] int? storyTagId, [FromQuery] int page = 0, [FromQuery] bool isAdmin = false)
		{
			var userIdClaim = HttpContext.User.FindFirstValue("userId");
			if (userIdClaim == null)
			{
				return BadRequest();
			}
			Expression<Func<Story, bool>> storyAuthorFilter = s => true;
			storyAuthorFilter = s => s.UserId == Int32.Parse(userIdClaim);
			if (teamsId != null)
			{
				storyAuthorFilter = s => s.TeamId == teamsId;
			} 
			else if (eventId != null)
			{
				storyAuthorFilter = s => s.EventId == eventId;
			}

			Expression<Func<Story, bool>> storyFilter = s => true;
			if (search != null)
			{
				storyFilter = s =>  s.Title.Contains(search);
			}

			Expression<Func<Story, bool>> storyTagsFilter = s => true;
			if (storyTagId.HasValue)
			{
				storyTagsFilter = s => s.StoryStoryTags.Any(sst => sst.StoryTagId == storyTagId.Value);
			}

            Expression<Func<Story, bool>> storyStatusFilter = s => s.Status != Status.ModerateAuto && s.Status != Status.ModerateByAdmin;
            if (isAdmin)
            {
                storyStatusFilter = s => s.Status == Status.Pending;
            }
            List<PublicApi_StoryModel> results = _storyRepository
                .GetAll()
                .Include(s => s.Pages)
                .Include(s => s.User)
                .Include(s => s.StoryVotes)
                .Include(s => s.StoryStoryTags)
                .ThenInclude(st => st.StoryTag)
                .Where(storyFilter)
                .Where(storyStatusFilter)
                .Where(storyTagsFilter)
                .Where(storyAuthorFilter)
                .Skip(5 * page)
                .Take(5)
                .Select(story => new PublicApi_StoryModel
                {
                    Id = story.Id,
                    Title = story.Title,
                    CoverUrl = story.CoverUrl,
                    Summary = story.Summary,
                    UserId = story.UserId,
                    TeamId = story.TeamId,
                    DateCreated = story.DateCreated,
                    Status = story.Status,
                    EventId = story.EventId,
                    HasVoted = story.StoryVotes.Any(sv => sv.UserId == Int32.Parse(userIdClaim)),
                    NbVote = story.StoryVotes.Count,
                    Pages = story.Pages.Select(page => new PublicApi_PageModel
                    {
                        Id = page.Id,
                        Content = page.Content,
                        LastUpdatedDateTime = page.LastUpdatedDateTime,
                        Order = page.Order,
                        StoryId = page.StoryId
                    }).ToList(),
                    StoryTags = story.StoryStoryTags.Select(sst => new PublicApi_StoryTagModel
                    {
                        Id = sst.StoryTag.Id,
                        Label = sst.StoryTag.Label,
                    }).ToList(),
                    User = new PublicApi_UserModel
                    {
                        AuthorName = story.User.AuthorName,
                        ProfilePictureUrl = story.User.ProfilePictureUrl
                    }
                }).ToList();

			return Ok(results);
		}

		[Authorize]
		[HttpGet("{storyId}")]
		public ActionResult Get([FromRoute] int storyId)
		{
			var userIdClaim = HttpContext.User.FindFirstValue("userId");
            Story story = _storyRepository.GetAll()
                .Include(s => s.User)
                .Include(s => s.StoryVotes)
                .Include(s => s.StoryStoryTags)
                .ThenInclude(st => st.StoryTag)
                .Include(s => s.Pages)
                .FirstOrDefault(s => s.Id == storyId);

                
			if (story == null)
			{
				return BadRequest();
			}

			return Ok(new PublicApi_StoryModel
			{
				Id = story.Id,
				Title = story.Title,
				CoverUrl = story.CoverUrl,
				Summary = story.Summary,
				UserId = story.UserId,
				TeamId = story.TeamId,
				DateCreated = story.DateCreated,
				Status = story.Status,
				EventId = story.EventId,
				HasVoted = story.StoryVotes.Any(sv => sv.UserId == Int32.Parse(userIdClaim)),
				NbVote = story.StoryVotes.Count,
				Pages = story.Pages.Select(page => new PublicApi_PageModel
				{
					Id = page.Id,
					Content = page.Content,
					LastUpdatedDateTime = page.LastUpdatedDateTime,
					Order = page.Order,
					StoryId = page.StoryId
				}).ToList(),
				StoryTags = story.StoryStoryTags.Select(sst => new PublicApi_StoryTagModel
				{
					Id = sst.StoryTag.Id,
					Label = sst.StoryTag.Label,
				}).ToList(),
				User = new PublicApi_UserModel
				{
					AuthorName = story.User.AuthorName,
					ProfilePictureUrl = story.User.ProfilePictureUrl,
				}
			});
		}

        [Authorize]
        [HttpGet("winner/{eventId}")]
        public ActionResult GetWinner([FromRoute] int eventId)
        {
            Story story = _storyRepository.GetAll().FirstOrDefault(s => s.EventId == eventId && s.Status == Status.Winner);
            if (story == null)
            {
                return BadRequest();
            }

            return Ok(new PublicApi_StoryModel
            {
                Id = story.Id,
                Title = story.Title,
                CoverUrl = story.CoverUrl,
                Summary = story.Summary,
                UserId = story.UserId,
                TeamId = story.TeamId,
                DateCreated = story.DateCreated,
                Status = story.Status,
                EventId = story.EventId
            });
        }
        
        [HttpPost]
		[Authorize]
		public async Task<IActionResult> CreateStory()
		{
			var userIdClaim = HttpContext.User.FindFirstValue("userId");
			
			try
			{
				var file = Request.Form.Files.GetFile("file");
				PublicApi_StoryModel storyToCreate = JsonConvert.DeserializeObject<PublicApi_StoryModel>(Request.Form["story"], MultiPartMessageJsonSerializerSettings);
				 
				if (file == null || file.Length == 0)
				{
					return BadRequest("Please select a file");
				}

				if (storyToCreate.EventId.HasValue)
				{
					Event eventLink = _eventRepository.GetById(storyToCreate.EventId.Value);
					if (eventLink != null)
					{
						DateTime today = DateTime.Now;
						if (eventLink.DateBegin <= today && eventLink.DateEnd >= today)
						{
							if (eventLink.Stories.Any(s => s.EventId == storyToCreate.EventId))
							{
								return BadRequest();
							}
						}
						else
						{
							return BadRequest();
						}
					}
					else
					{

						return NotFound();
					}
				}
				//Todo si y a un event envoyé il faut vérifier qu'on a bien encore le droit de participer date + duplicité.
				using (var stream = new MemoryStream())
				{
					await file.CopyToAsync(stream);
					using (var doc = WordprocessingDocument.Open(stream, false))
					{
                        var body = doc.MainDocumentPart.Document.Body;
                        string text = "";
                        foreach (var paragraph in body.Elements<Paragraph>())
                        {
                            foreach (var run in paragraph.Elements<Run>())
                            {
                                text += run.InnerText + " ";
                            }
                            text += Environment.NewLine; // Ajoute un retour à la ligne après chaque paragraphe
                        }
                        
                        StoryHelper storyHelper = new StoryHelper();
						List<string> pages = storyHelper.SplitIntoPages(text, 250);

                        Story story = new Story
						{
							Title = storyToCreate.Title,
							CoverUrl = storyToCreate.CoverUrl,
							UserId = Int32.Parse(userIdClaim),
							DateCreated = DateTime.Now,
							Status = Status.Pending,
							EventId = storyToCreate.EventId,
						};
						_storyRepository.Add(story);
						_storyRepository.Save();
						int orderPage = 0;
						foreach (string page in pages)
						{
							Page pageToCreate = new Page
							{
								Content = page,
								LastUpdatedDateTime = DateTime.Now,
								Order = orderPage,
								StoryId = story.Id
							};
							_pageRepository.Add(pageToCreate);
							_pageRepository.Save();
                            orderPage++;
						}

						if (storyToCreate.StoryTags != null)
						{
							foreach (PublicApi_StoryTagModel storyTagToLink in storyToCreate.StoryTags)
							{
								StoryStoryTag storyStoryTag = new StoryStoryTag
								{
									StoryId = story.Id,
									StoryTagId = storyTagToLink.Id
								};
								_storyStoryTagRepository.Add(storyStoryTag);
								_storyStoryTagRepository.Save();
							}
						}
					}
				}

				return Ok(new { message = "File uploaded successfully" });
			}
			catch (Exception ex)
			{
				return StatusCode(StatusCodes.Status500InternalServerError, "Error uploading file");
			}
		}


        [HttpPut]
        public async Task<IActionResult> Update([FromBody] PublicApi_StoryModel story)
        {
            var userIdClaim = HttpContext.User.FindFirstValue("userId");
            bool fieldhasBeenChange = false;
            Story storyToUpdate = _storyRepository.GetById(story.Id);
            if (storyToUpdate == null)
            {
                return NotFound();
            }

            if (story.Status != storyToUpdate.Status)
            {
                storyToUpdate.Status = story.Status;
                fieldhasBeenChange = true;
            }

            if (fieldhasBeenChange)
            {
				_storyRepository.Update(storyToUpdate);
				_storyRepository.Save();
            }

            return Ok();
        }
    }
}

File: lovecraft-ts/src/controllers/v1/storyController.ts
import { MultipartFile, MultipartValue } from '@fastify/multipart'
import { Status } from '@prisma/client'
import { FastifyReply, FastifyRequest } from 'fastify'
import { STANDARD } from '../../constants/request'
import { ERRORS, handleServerError } from '../../helpers/errors.helper'
import { extractRawTextFromDocx, splitIntoPages, toStoryDto } from '../../helpers/story.helper'
import { IStoryDto, StoryWithRelations } from "../../schemas/Story"
import { prisma } from '../../utils'

export const storyController = {
  getAll: async (request: FastifyRequest, reply: FastifyReply) => {
    try {
      const { userId, teamsId, eventId, search, storyTagId, page = 0, order = 'desc', mode } = request.query as {
        userId?: string
        teamsId?: number
        eventId?: number
        search?: string
        storyTagId?: number
        page?: number
        order?: string
        mode?: string
      }
      const currentUserId = request['authUser'].id
      if (!currentUserId) {
        return reply
          .code(ERRORS.userNotExists.statusCode)
          .send(ERRORS.userNotExists.message)
      }

      const filters: any = {}

      if (teamsId) {
        filters.teamId = teamsId
        filters.status = 'Online'
      } else if (eventId) {
        filters.eventId = eventId
        filters.status = 'Online'
      } else if (userId && userId !== '') {
        const userIdNumber = Number(userId)
        filters.userId = userIdNumber
        if (userIdNumber === currentUserId) {
          if (mode === 'draft') {
            filters.status = 'Pending'
          }
        } else {
          filters.status = 'Online'
        }
      } else {
        filters.status = 'Online'
      }

      if (search) {
        filters.title = { contains: search, mode: 'insensitive' }
      }

      if (storyTagId) {
        filters.StoryStoryTags = {
          some: { storyTagId }
        }
      }

      const orderBy = { createdAt: order === 'asc' ? 'asc' as const : 'desc' as const }

      const stories: StoryWithRelations[] = await prisma.story.findMany({
        where: filters,
        include: {
          pages: true,
          user: true,
          team: true,
          event: true,
          storyVotes: true,
          storyStoryTags: {
            include: { storyTag: true }
          },
          storyHistories: {
            where: { userId: currentUserId },
            take: 1
          },
          _count: {
            select: {
              storyComments: true
            }
          }
        },
        skip: page * 5,
        take: 5,
        orderBy
      })
      const results = stories.map(story => { return toStoryDto(story) })
      return reply.code(STANDARD.OK.statusCode).send(results)
    } catch (error) {
      console.error(error)
      return handleServerError(reply, error)
    }
  },
  getById: async (
    request: FastifyRequest<{ Params: { id: string } }>,
    reply: FastifyReply
  ) => {
    const storyId = Number(request.params.id)
    const currentUserId = request['authUser'].id
    if (!currentUserId) {
      return reply
        .code(ERRORS.userNotExists.statusCode)
        .send(ERRORS.userNotExists.message)
    }
    const story: StoryWithRelations = await prisma.story.findUnique({
      where: { id: storyId },
      include: {
        pages: true,
        user: true,
        team: true,
        event: true,
        storyVotes: true,
        storyStoryTags: {
          include: { storyTag: true }
        },
        storyHistories: {
          where: { userId: currentUserId },
          take: 1
        },
        _count: {
          select: {
            storyComments: true
          }
        }
      },
    })
    if (story == null) {
      return reply
        .code(ERRORS.NotFound.statusCode)
        .send(ERRORS.NotFound.message)
    }
    return reply.code(STANDARD.OK.statusCode).send(toStoryDto(story))
  },
  post: async (
    request: FastifyRequest<{ Body: { file: any; story: string } }>,
    reply: FastifyReply,
  ) => {
    try {
      const currentUserId = request['authUser']?.id
      if (!currentUserId) {
        return reply
          .code(ERRORS.userNotExists.statusCode)
          .send(ERRORS.userNotExists.message)
      }
      const parts = await request.parts()

      let fileBuffer: Buffer | null = null
      let fileName: string | null = null
      let mimeType: string | null = null
      let storyToCreate: IStoryDto | null = null
      let docxText: string
      for await (const part of parts) {
        if (part.type === 'file') {
          const filePart = part as MultipartFile
          fileBuffer = await filePart.toBuffer()
          fileName = filePart.filename
          mimeType = filePart.mimetype
          docxText = await extractRawTextFromDocx(fileBuffer)
        } else if (part.type === 'field') {
          const valuePart = part as MultipartValue
          if (valuePart.fieldname === 'story') {
            storyToCreate = JSON.parse(valuePart.value as string)
          }
        }
      }

      if (storyToCreate.eventId) {
        const eventLink = await prisma.event.findUnique({
          where: { id: storyToCreate.eventId }
        })
        if (!eventLink) {
          return reply.code(ERRORS.NotFound.statusCode).send('Event not found')
        }

        const today = new Date()
        if (eventLink.dateBegin > today || eventLink.dateEnd < today) {
          return reply.code(ERRORS.badRequest.statusCode).send('Event is not active')
        }

        const existingStory = await prisma.story.findFirst({
          where: { userId: currentUserId, eventId: storyToCreate.eventId }
        })
        if (existingStory) {
          return reply.code(ERRORS.badRequest.statusCode).send('Already participated in this event')
        }
      }

      const pages = await splitIntoPages(docxText, 250) // 250 mots par page

      const createdStory = await prisma.story.create({
        data: {
          title: storyToCreate.title,
          coverUrl: storyToCreate.coverUrl,
          userId: currentUserId,
          status: Status.Pending,
          eventId: storyToCreate.eventId
        }
      })

      const pagesData = pages.map((content, index) => ({
        storyId: createdStory.id,
        content,
        order: index,
        createdAt: new Date(),
        updatedAt: new Date()
      }))
      await prisma.page.createMany({ data: pagesData })

      if (storyToCreate.storyTags?.length) {
        const tagsData = storyToCreate.storyTags.map((tag: any) => ({
          storyId: createdStory.id,
          storyTagId: tag.id
        }))
        await prisma.storyStoryTag.createMany({ data: tagsData })
      }

      const finalStory: StoryWithRelations = await prisma.story.findUnique({
        include: {
          pages: true,
          user: true,
          team: true,
          event: true,
          storyVotes: true,
          storyStoryTags: {
            include: { storyTag: true }
          },
          storyHistories: {
            where: { userId: currentUserId },
            take: 1
          },
          _count: {
            select: {
              storyComments: true
            }
          }
        },
        where: { id: createdStory.id }
      })
      return reply.code(STANDARD.OK.statusCode).send(toStoryDto(finalStory))
    } catch (error) {
      console.error(error)
      return handleServerError(reply, error)
    }
  }
}